common/src/main/java/noobanidus/mods/lootr/common/block/entity/BlockEntityTicker.java

```diff
@@ -13,187 +12,187 @@
-      return;
+      entityPositions.add(validEntity.getBlockPos());
     }
 
-    ChunkPos chunkPos = new ChunkPos(position);
-
-    Set<ChunkPos> chunks = new ObjectOpenHashSet<>();
-    chunks.add(chunkPos);
-
-    int oX = chunkPos.x;
-    int oZ = chunkPos.z;
-
-    for (int x = -2; x <= 2; x++) {
-      for (int z = -2; z <= 2; z++) {
-        ChunkPos newPos = new ChunkPos(oX + x, oZ + z);
-        // This has the potential to force-load chunks on the main thread
-        // by ignoring the loading state of chunks outside the world border.
-        if (!LootrAPI.isWorldBorderSafe(level, newPos)) {
-          continue;
-        }
-
-        chunks.add(newPos);
+    if (!entityPositions.isEmpty()) {
+      Entry entry = new Entry(dimension, chunkPos, entityPositions);
+      synchronized (pendingEntries) {
+        pendingEntries.add(entry);
       }
     }
-
-    Entry newEntry = new Entry(dimension, position, chunks, LootrAPI.getCurrentTicks());
-    synchronized (pendingEntries) {
-      pendingEntries.add(newEntry);
-    }
   }
 
-  public static void onServerTick() {
+  public static void onServerTick(MinecraftServer server) {
     if (LootrAPI.isDisabled()) {
       return;
     }
-    MinecraftServer server = LootrAPI.getServer();
-    if (server == null) {
-      LootrAPI.LOG.error("MinecraftServer was null during ServerTickEvent!");
-      return;
-    }
+
     Iterator<Entry> iterator = blockEntityEntries.iterator();
     while (iterator.hasNext()) {
       Entry entry = iterator.next();
-      ServerLevel level = server.getLevel(entry.getDimension());
-      if (level == null || LootrAPI.hasExpired(entry.age(server)) || (!LootrAPI.isWorldBorderSafe(level, entry.getPosition()))) {
-        iterator.remove();
-        continue;
-      }
 
-      // There's a significant level of "duplication" here:
-      // This check consults the actual chunk source.
-      if (!level.getChunkSource().hasChunk(entry.getPosition().getX() >> 4, entry.getPosition().getZ() >> 4)) {
+      ServerLevel level = server.getLevel(entry.dimension());
+      if (level == null) {
+        iterator.remove();
         continue;
       }
 
-      // And for all chunks in a 5x5 area. Modifying a block entity or block
-      // will cause it to propagate ticks, etc, which can cause chunk loading.
-      boolean skip = false;
-      for (ChunkPos chunkPos : entry.getChunkPositions()) {
-        if (!level.getChunkSource().hasChunk(chunkPos.x, chunkPos.z)) {
-          skip = true;
-          break;
+      switch (entry.getChunkLoadStatus(level)) {
+        case UNLOADED -> {
+          // the chunk has unloaded. this entry is no longer valid, and it will be added again if the chunk loads again.
+          iterator.remove();
+        }
+        case SURROUNDING_CHUNKS_NOT_LOADED -> {
+          // keep waiting for the surrounding chunks to load
+        }
+        case FULLY_LOADED -> {
+          replaceEntitiesInChunk(level, entry);
+          iterator.remove();
```

Comment:
I was initially concerned that there might be issues where a loot table hasn't been set by the time this is processed, but I can't think of any situations where an eligible block entity would be created and not have a loot table assigned to it by the time the chunk is actually promoted.